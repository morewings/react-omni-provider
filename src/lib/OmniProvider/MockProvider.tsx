import type {FC, ReactNode} from 'react';

export const MockProvider: FC<{children: ReactNode; testId?: string; color?: string}> = ({children, testId, color}) => {
    return (
        <div data-testid={testId} style={{padding: '36px', backgroundColor: color}}>
            {children}
        </div>
    );
};
