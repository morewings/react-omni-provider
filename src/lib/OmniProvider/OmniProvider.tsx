import type {FC, ReactNode} from 'react';

import type {ProviderConfig} from './ProviderConfig.ts';

type OmniProviderProps = {
    children: ReactNode;
    providerConfig?: ProviderConfig[];
};

export const OmniProvider: FC<OmniProviderProps> = ({children, providerConfig = []}) => {
    return providerConfig.reverse().reduce((prev, current) => {
        if (Array.isArray(current)) {
            const [Provider, props] = current;
            return <Provider {...props}>{prev}</Provider>;
        } else {
            const Provider = current;
            return <Provider>{prev}</Provider>;
        }
    }, children);
};
