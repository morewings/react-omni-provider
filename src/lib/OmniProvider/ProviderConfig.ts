import type {ComponentType, ComponentProps} from 'react';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ProviderConfig<TComponent extends ComponentType<any> = ComponentType<any>> =
    | TComponent
    | [TComponent, ComponentProps<TComponent>];
