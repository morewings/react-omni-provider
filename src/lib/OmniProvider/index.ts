export {OmniProvider} from './OmniProvider';
export {withOmniProvider} from './withOmniProvider.tsx';
