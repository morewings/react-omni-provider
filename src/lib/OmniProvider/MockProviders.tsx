import type {ComponentProps} from 'react';

import type {ProviderConfig} from './ProviderConfig.ts';
import {MockProvider} from './MockProvider.tsx';

export const mockProviders = [
    [MockProvider, {testId: 'foo'} as ComponentProps<typeof MockProvider>],
    [MockProvider, {testId: 'bar'} as ComponentProps<typeof MockProvider>],
    [MockProvider, {testId: 'bazz'} as ComponentProps<typeof MockProvider>],
] as ProviderConfig[];
