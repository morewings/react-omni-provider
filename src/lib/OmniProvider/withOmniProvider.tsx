import type {ComponentType, FC, JSX} from 'react';

import {OmniProvider} from './OmniProvider.tsx';
import type {ProviderConfig} from './ProviderConfig.ts';

export const withOmniProvider =
    (providers: ProviderConfig[]) =>
    <TProps extends JSX.IntrinsicAttributes>(Component: ComponentType<TProps>) => {
        const result: FC<TProps> = props => (
            <OmniProvider providerConfig={providers}>
                <Component {...props} />
            </OmniProvider>
        );
        result.displayName = 'OmniProvider';
        return result;
    };
