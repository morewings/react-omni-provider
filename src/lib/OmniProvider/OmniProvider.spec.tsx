import {render} from '@testing-library/react';
import type {ComponentProps} from 'react';

import {OmniProvider} from './OmniProvider.tsx';
import {MockProvider} from './MockProvider.tsx';

describe('lib > OmniProvider', () => {
    it('renders children', () => {
        const children = 'foo';
        const {getByText} = render(<OmniProvider>{children}</OmniProvider>);
        expect(getByText(children)).toBeInTheDocument();
    });

    it('renders given provider', () => {
        const id = 'foo';
        const {getByTestId} = render(
            <OmniProvider providerConfig={[[MockProvider, {testId: id} as ComponentProps<typeof MockProvider>]]}>
                foo
            </OmniProvider>
        );
        expect(getByTestId(id)).toBeInTheDocument();
    });

    it('respects left to right order of providers given', () => {
        const {asFragment} = render(
            <OmniProvider
                providerConfig={[
                    [MockProvider, {testId: 'foo'} as ComponentProps<typeof MockProvider>],
                    [MockProvider, {testId: 'bar'} as ComponentProps<typeof MockProvider>],
                    [MockProvider, {testId: 'bazz'} as ComponentProps<typeof MockProvider>],
                ]}>
                foo
            </OmniProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
