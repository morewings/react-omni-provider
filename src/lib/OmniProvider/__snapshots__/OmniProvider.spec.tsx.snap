// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib > OmniProvider respects left to right order of providers given 1`] = `
<DocumentFragment>
  <div
    data-testid="foo"
    style="padding: 36px;"
  >
    <div
      data-testid="bar"
      style="padding: 36px;"
    >
      <div
        data-testid="bazz"
        style="padding: 36px;"
      >
        foo
      </div>
    </div>
  </div>
</DocumentFragment>
`;
